faunadb-common/src/main/java/com/faunadb/common/Connection.java

```diff
@@ -358,90 +352,82 @@ public void syncLastTxnTime(long newTxnTime) {
-    FullHttpRequest request = newRequest(HttpMethod.PATCH, path, body);
-    return performRequest(request, queryTimeout);
-  }
-
-  private FullHttpRequest newRequest(HttpMethod method, String path) throws IOException {
-    return new DefaultFullHttpRequest(HttpVersion.HTTP_1_1, method, mkUrl(path));
+  public CompletableFuture<HttpResponse<String>> patch(String path, JsonNode body, Optional<Duration> queryTimeout) throws Exception {
+    return performRequest("PATCH", path, Optional.of(body), Map.of(), queryTimeout);
   }
 
-  private void fixRequestParameters(FullHttpRequest request, Map<String, List<String>> params) {
-    QueryStringEncoder encoder = new QueryStringEncoder(request.uri());
-
-    for (Map.Entry<String, List<String>> entries : params.entrySet()) {
-      String k = entries.getKey();
-      for (String v : entries.getValue())
-        encoder.addParam(k, v);
-    }
-
-    request.setUri(encoder.toString());
-  }
-
-  private FullHttpRequest newRequest(HttpMethod method, String path, JsonNode body) throws IOException {
-    FullHttpRequest request = newRequest(method, path);
-
-    byte[] jsonBody = json.writeValueAsBytes(body);
-    request.content().clear().writeBytes(jsonBody);
-
-    request.headers().set(HttpHeaderNames.CONTENT_LENGTH, jsonBody.length);
-    request.headers().set(HttpHeaderNames.CONTENT_TYPE, "application/json; charset=utf-8");
-
-    return request;
+  public static URI appendUri(URI oldUri, String queryKey, String queryValue) throws URISyntaxException {
+    String urlEncodedKey = URLEncoder.encode(queryKey, StandardCharsets.UTF_8);
+    String urlEncodedValue = URLEncoder.encode(queryValue, StandardCharsets.UTF_8);
+    String query = urlEncodedKey + "=" + urlEncodedValue;
+    return new URI(oldUri.getScheme(), oldUri.getAuthority(), oldUri.getPath(),
+            oldUri.getQuery() == null ? query : oldUri.getQuery() + "&" + query, oldUri.getFragment());
   }
 
-  private CompletableFuture<FullHttpResponse> performRequest(final FullHttpRequest request, final Optional<Duration> requestQueryTimeout) {
+  private CompletableFuture<HttpResponse<String>> performRequest(String httpMethod, String path, Optional<JsonNode> body,
+                                                                 Map<String, String> params, final Optional<Duration> requestQueryTimeout) throws Exception {
     final Timer.Context ctx = registry.timer("fauna-request").time();
-    final CompletableFuture<FullHttpResponse> rv = new CompletableFuture<>();
+    final CompletableFuture<HttpResponse<String>> rv = new CompletableFuture<>();
 
-    request.headers().add("Authorization", authHeader);
-    request.headers().set("X-FaunaDB-API-Version", API_VERSION);
+    URI requestUri = URI.create(mkUrl(path));
 
-    if(jvmDriver != null) {
-      request.headers().set(X_FAUNA_DRIVER, jvmDriver.toString());
+    // Encode all query parameters
+    for (Map.Entry<String,String> entry : params.entrySet()) {
```

Comment:
FYI in my streaming PRs I have re-promoted the parameters to a `MultiMap`.

Reply:
The changes can be applied in the streaming PR, right? In other words, we don't need them to include here as well, right?

Reply:
AFAIK there is not need for it in this PR indeed